### PR TITLE
Board support : Pixhawk Pro

### PR DESCRIPTION
**Pull request by klopezal**
This pull request brings changes in order to use ArduPilot on Pixhawk Pro hardware. As this board uses a different processor (with more flash and RAM, but also different register definitions), it also needs subsequent changes in PX4NuttX and PX4Firmware : these two pull requests should be tested and validated before considering merging : 

https://github.com/ArduPilot/PX4NuttX/pull/11
https://github.com/ArduPilot/PX4Firmware/pull/76

### CODE REVIEW

File: libraries/AP_BoardConfig/AP_BoardConfig.h [hunk 1]
```diff
@@ -63,6 +63,7 @@ class AP_BoardConfig {
         PX4_BOARD_PIXRACER = 4,
         PX4_BOARD_PHMINI   = 5,
         PX4_BOARD_PH2SLIM  = 6,
+        PX4_BOARD_PIXHAWK_PRO  = 7,
```
Comment (magicrub) [inline libraries/AP_BoardConfig/AP_BoardConfig.h]: I thought 7 was used somewhere hidden which is why AEROFC moved up to 13? Need to double-check this.

File: libraries/AP_Compass/AP_Compass.cpp [hunk 2]
```diff
@@ -607,6 +608,17 @@ void Compass::_detect_backends(void)
...
+
+    case AP_BoardConfig::PX4_BOARD_PIXHAWK_PRO:
+//        ADD_BACKEND(AP_Compass_HMC5843::probe(*this, hal.spi->get_device(HAL_COMPASS_HMC5843_NAME),
+//                                               false, ROTATION_PITCH_180),
+//                     AP_Compass_HMC5843::name, false);
+        ADD_BACKEND(AP_Compass_AK8963::probe_mpu9250(*this, 0, ROTATION_ROLL_180_YAW_90),
+                     AP_Compass_AK8963::name, false);
+//		ADD_BACKEND(AP_Compass_LIS3MDL::probe(*this, hal.spi->get_device(HAL_COMPASS_LIS3MDL_NAME),
+//                                               false, ROTATION_PITCH_180),
+//                     AP_Compass_LIS3MDL::name, false);
```
Comment (magicrub) [inline libraries/AP_Compass/AP_Compass.cpp]: as a rule, we don't like to add code that is commented out

File: libraries/AP_BoardConfig/AP_BoardConfig.h (hunk 1 above)
Comment (OXINARF) [inline libraries/AP_BoardConfig/AP_BoardConfig.h]: Yes, 7 to 12 is occupied with VRBrain boards.

Comment (klopezal) [inline libraries/AP_BoardConfig/AP_BoardConfig.h]: That is now corrected, thanks!

File: libraries/AP_Compass/AP_Compass.cpp (hunk 2 above)
Comment (klopezal) [inline libraries/AP_Compass/AP_Compass.cpp]: Also corrected.

**Comment by tridge** [comment]
Thank you very much for this, you've saved me a lot of time!
I've pushed the PX4NuttX and PX4Firmware changes to pr-ppro branches on the ArduPilot repos. That will allow travis to work on them.
I've also pushed the ArduPilot changes, rebased on master, to https://github.com/tridge/ardupilot/commits/pr-ppro
I tested it on a ppro here (thank you for sending me one btw!). It does build and load fine, and boots up, but there are a few issues I've noticed so far:

- power status shows USB, Brick and servo power, when it only has USB power
- primary mag seems ok, 2nd mag is giving crazy output
- both accels and gyros are working
- CPU load looks fine

I haven't tried external i2c or any other peripherals yet. Great start though, thanks!

**Comment by tridge** [comment]
we should also add this to travis

**Comment by klopezal** [comment]
Thanks! 
- I spotted the power error, it is in PX4Firmware, I'll suggest a correction
- true, SPI communication with LIS3MDL does not work well for the moment, I always get 0x00 or 0xFF bytes on the bus, pending investigation
- ok
- ok

**Comment by klopezal** [comment]
@tridge There was an issue with DMA configuration that is now solved. LIS3MDL now gives correct output. Right now the SPI bus is used at 1 MHz, we could go up to  10 MHz according to the datasheet, I haven't tested yet.

**Comment by klopezal** [comment]
LIS3MDL now works correctly, I had to decrease SPI speed down to 500 KHz, otherwise there are constant communication errors on the first bit of OUT_XXX_L registers.

The LIS3MDL sensor has an intrinsic bigger offset on Z axis, hence the MAX_OFFSET modification. Should we make this change specific to this sensor or let the modification for all compasses?

**Comment by tridge** [comment]
I don't mind increasing that default to 850.
glad to hear the lis3mdl is working now!

**Comment by tridge** [comment]
I've done some commit msg cleanup to follow our coding guidelines, plus pushed the PX4Firmware and PX4NuttX changes
the commit cleanups are here:
https://github.com/tridge/ardupilot/tree/ppro-wip

**Comment by tridge** [comment]
looks like it breaks the build for px4-v1:
armv7-m/up_vectors.c:50:4: error: #error ARMV7M_PERIPHERAL_INTERRUPTS must be defined to the number of I/O interrupts to be supported

 #  error ARMV7M_PERIPHERAL_INTERRUPTS must be defined to the number of I/O interrupts to be supported

**Comment by tridge** [comment]
I've updated your PR branch with a patch to PX4NuttX that fixes the FMUv1 build. Note that I only fixed the build, I haven't yet tested the build and run on FMUv1, FMUv2 and FMUv4. We'll need to carefully test all boards before merging.
@klopezal please look at the change I made to PX4NuttX here:
https://github.com/ArduPilot/PX4NuttX/commit/ea84057ec705a39b67947e690a9dfd79df9d5017
the problem was the 405RG wasn't being handled. Previously a lot of the ifdef blocks had a default else, but with the ppro changes they don't. That is what broken FMUv1

**Comment by tridge** [comment]
I've now done basic testing on FMUv1, FMUv2, FMUv4 and the v4pro
The only failure I'm seeing is the IO firmware on FMUv1. That needs some more work.

**Comment by tridge** [comment]
I have found the bug with px4io on fmuv1. It is an existing bug, not related to this PR. It is caused by the way px4io update uses a uart instead of i2c for the update, and that uart is already registered in HAL_PX4. That is unique to the v1 boards. 
I'll fix the bug soon.

**Comment by tridge** [comment]
this PR fixes the FMUv1 problem:
https://github.com/ArduPilot/ardupilot/pull/6645
once that is merged I think we'll be ready to add this v4pro PR

**Comment by tridge** [comment]
this has now been merged into master
@klopezal many thanks for your work on this!

**Pull request closed without merging**